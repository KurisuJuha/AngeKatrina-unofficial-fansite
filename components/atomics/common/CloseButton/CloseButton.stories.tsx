import { Meta, Story } from "@storybook/react/types-6-0";
import CloseButton, { CloseButtonProps } from "./CloseButton";

export default {
  title: "common/CloseButton",
  component: CloseButton,
  args: {
    displayAnimationDelay: 500,
    disableAnimationDelay: 500,
    runDisplayAnimation: true,
    runCloseAnimation: false,
  },
} as Meta<CloseButtonProps>;

const Template: Story<CloseButtonProps> = (args) => <CloseButton {...args} />;

export const NormalClose = Template.bind({});
