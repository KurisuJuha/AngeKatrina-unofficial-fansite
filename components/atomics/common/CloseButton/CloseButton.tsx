import React, { useState } from "react";
import styled, { css } from "styled-components";
import { ANGE_WHITE, ANGE_LIVE_BACK_COLOR } from "../../../../constants/colors";
import { boundExpand } from "../../../../styles/commonAnimation";

interface Props {
  displayAnimationDelay: number;
  onClickFC?: () => void;
  top?: string;
  right?: string;
  bottom?: string;
  left?: string;
}

const Wrapper = styled.div<Required<Omit<Props, "onClickFC">>>`
  position: absolute;
  top: ${({ top }) => top};
  right: ${({ right }) => right};
  bottom: ${({ bottom }) => bottom};
  left: ${({ left }) => left};
  display: flex;
  justify-content: center;
  align-items: center;
  width: 55px;
  height: 55px;
  animation: ${boundExpand(40, 1.1)} 500ms ease-out both
    ${({ displayAnimationDelay }) => displayAnimationDelay}ms;
`;

const transitionTime = "200ms";

const CloseButtonOutLine = styled.button`
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  background-color: ${ANGE_WHITE};
  border: 2px solid ${ANGE_LIVE_BACK_COLOR};
  cursor: pointer;
  border-radius: 50%;
  transition: background-color ${transitionTime}, border ${transitionTime};
  :hover {
    background-color: ${ANGE_LIVE_BACK_COLOR};
    border: 2px solid ${ANGE_WHITE};
  }
`;

const CloseButtonMain = styled.div<{ isHovering: boolean }>`
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  width: calc(100% - 4px);
  height: calc(100% - 4px);
  background-color: ${({ isHovering }) =>
    isHovering ? ANGE_WHITE : ANGE_LIVE_BACK_COLOR};
  transition: background-color ${transitionTime};
  border-radius: 50%;
`;

const lineCSS = css<{ isHovering: boolean }>`
  position: absolute;
  display: flex;
  height: 5px;
  width: 70%;
  border-radius: 3px;
  transition: background-color ${transitionTime};
  background-color: ${({ isHovering }) =>
    isHovering ? ANGE_LIVE_BACK_COLOR : ANGE_WHITE};
`;

const LeftTopLine = styled.div<{ isHovering: boolean }>`
  ${lineCSS}
  transform: rotate(-45deg);
`;

const RightTopLine = styled.div<{ isHovering: boolean }>`
  ${lineCSS}
  transform: rotate(45deg);
`;

const CloseButton: React.FC<Props> = ({
  onClickFC,
  top = "auto",
  right = "auto",
  bottom = "auto",
  left = "auto",
  displayAnimationDelay,
}: Props) => {
  const [isHovering, changeIsHovering] = useState(false);

  return (
    <Wrapper
      top={top}
      right={right}
      bottom={bottom}
      left={left}
      displayAnimationDelay={displayAnimationDelay}
    >
      <CloseButtonOutLine
        onClick={onClickFC}
        onMouseEnter={() => changeIsHovering(true)}
        onMouseLeave={() => changeIsHovering(false)}
      >
        <CloseButtonMain isHovering={isHovering}>
          <LeftTopLine isHovering={isHovering} />
          <RightTopLine isHovering={isHovering} />
        </CloseButtonMain>
      </CloseButtonOutLine>
    </Wrapper>
  );
};

export default CloseButton;
