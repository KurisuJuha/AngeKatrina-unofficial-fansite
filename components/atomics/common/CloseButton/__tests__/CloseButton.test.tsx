import React from "react";
import CloseButton from "../CloseButton";
import renderer from "react-test-renderer";

test("sample", () => {
  const component = renderer.create(<CloseButton />);
  const tree = component.toJSON() as renderer.ReactTestRendererJSON;
  expect(tree).toMatchSnapshot();
});
