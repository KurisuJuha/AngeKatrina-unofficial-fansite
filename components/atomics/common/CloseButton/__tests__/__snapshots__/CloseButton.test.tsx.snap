// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sample 1`] = `
<div
  className="CloseButton__Wrapper-rlgiig-0 dOYOCz"
>
  <button
    className="CloseButton__CloseButtonOutLine-rlgiig-1 fKjHHz"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="CloseButton__CloseButtonMain-rlgiig-2 bGdhyS"
    >
      <div
        className="CloseButton__LeftTopLine-rlgiig-3 deprwd"
      />
      <div
        className="CloseButton__RightTopLine-rlgiig-4 dgfLAc"
      />
    </div>
  </button>
</div>
`;
